File: ase-backend/src/main/java/de/corvinrose/ase/service/LikeService.java
package de.corvinrose.ase.service;

import de.corvinrose.ase.model.Like;
import de.corvinrose.ase.repository.LikeRepository;
import java.util.List;
import java.util.UUID;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@AllArgsConstructor
public class LikeService {

    private LikeRepository likeRepository;

    public List<Like> findAllLikes() {
        return likeRepository.findAll();
    }

    public List<UUID> findAllLikesByShader(UUID shaderId) {
        return likeRepository.findAllLikesByShader(shaderId);
    }

    public List<UUID> findAllLikesByUser(UUID userId) {
        return likeRepository.findAllLikesByUser(userId);
    }

    public Like addLike(Like like) {
        return likeRepository.save(like);
    }

    public void deleteLike(Like like) {
        likeRepository.deleteLikeById(like.getShaderId(), like.getUserId());
    }
}

File: backend/Core/App/Service/LikeService.php
<?php

namespace Core\App\Service;

use Core\App\Provider\DBConnection;
use Core\Domain\Model\LikeModel;
use Exception;

class LikeService {
    private DBConnection $dbConnection;

    public function __construct() {
        $this->dbConnection = new DBConnection();
    }

    /** @throws Exception */
    public function findAllLikes(): array {
        $likes = $this->dbConnection->runSql("SELECT * FROM ase_likes");
        return array_map(fn($v) => LikeModel::fromDBObject($v), $likes);
    }

    /** @throws Exception */
    public function findAllLikesByShader($id = "-1"): array {
        $likes = $this->dbConnection->runSql("SELECT u.id FROM ase_likes l, ase_user u 
                          WHERE l.shader_id = '$id' AND u.id = l.user_id");
        return array_map(fn($v) => $v["id"], $likes);
    }

    /** @throws Exception */
    public function findAllLikesByUser($id = "-1"): array {
        $likes = $this->dbConnection->runSql("SELECT s.id FROM ase_likes l, ase_shader s 
                          WHERE s.author_id = '$id' AND s.id = l.shader_id");
        return array_map(fn($v) => $v["id"], $likes);
    }

    /** @throws Exception */
    public function addLike($like = null) {
        $this->dbConnection->runSql("INSERT INTO ase_likes (shader_id, user_id)
                 VALUES ('$like->shaderId', '$like->userId')");
        return $like;
    }

    /** @throws Exception */
    public function deleteLike($like = null) {
        $this->dbConnection->runSql(
            "DELETE FROM ase_likes WHERE shader_id='$like->shaderId' AND user_id='$like->userId'"
        );
        return null;
    }
}
